Mergify/bp/release/v3.x/pr 1663

jeoghyukLee123 opened the pull request:
Draft PR for upgrade handler / epoching spam prevention

dongsam commented on go.mod:
```diff
@@ -233,9 +233,11 @@ require (
 	github.com/tidwall/btree v1.7.0 // indirect
 	github.com/tyler-smith/go-bip39 v1.1.0 // indirect
 	github.com/ulikunitz/xz v0.5.11 // indirect
+	github.com/wealdtech/go-eth2-wallet-encryptor-keystorev4 v1.4.1
```
It seems that [Prysm is using keystore v1.1.3](https://github.com/prysmaticlabs/prysm/blob/v5.2.0/go.mod#L84). Could you clarify why you chose v1.4.1? It’s important to clearly understand the differences between the two versions to make an informed decision.

wnjoon replied:
It seems that the latest version was automatically imported when the go get command was executed.

I changed go.mod file in commit [dda705f](https://github.com/b-harvest/babylon/pull/2/commits/dda705f6008a791b7086878278f26d9157ce7d8d)

jeoghyukLee123 committed:
feat: epoching spam prevention (#1663)

# Description
Closes: #1662

## Objective
To mitigate spamming attacks against the Epoching module, this PR
introduces the following:

### Fund locking at message enqueue

Lock funds when a message is enqueued.

**Messages subject to fund locking:**

- Checkpointing module: MsgWrappedCreateValidator

- Epoching module: MsgWrappedDelegate

### Mandatory execution gas charge for end-of-epoch messages

**Messages subject to the mandatory execution gas charge:**

- Checkpointing module: MsgWrappedCreateValidator

- Epoching module: MsgWrappedEditValidator, MsgWrappedDelegate,
MsgWrappedUndelegate, MsgWrappedBeginRedelegate,
MsgWrappedCancelUnbondingDelegation

### Minimum amount requirement at enqueue time
Enforce a minimum amount when enqueuing messages to prevent DoS attacks.
## Note
- Add delegation_pool.go : The module manages fund lock / unlock related
works
- Add delegation_pool_test.go : unit test for fund lock / unlock
- Add module parameters for execution gas
- Apply GasMeter().ConsumeGas in epoching / checkpointing module tx
- Add unit test for execution gas
- Modify Endblock of epoching module : When epoch ends, it unlocks fund
and executes the message

## Progress check list

- [x] implement fund lock / unlock
- [x] implement adding module parameters for gas for execution
- [x] implement migration for adding module params
- [x] unit test for fund lock
- [x] unit test for gas consumption
- [x] e2e test (WrappedDelegate / WrappedCreateValidator)
- [ ] estimate worst execution gas

## how to review
`make test-e2e-cache-epoching-spam-prevention`

---------

Co-authored-by: jhlee <jhlee@jhleeui-MacBookAir.local>
(cherry picked from commit bad5e6411f683fb978ae662d830cfd70d676f4ca)

# Conflicts:
#	CHANGELOG.md
#	app/app.go
#	app/app_test.go
#	client/docs/swagger-ui/swagger.yaml
#	x/epoching/types/events.pb.go
#	x/epoching/types/params.pb.go

GAtom22 committed:
fix conflicts

Commit:
add upgrades.go draft

Commit:
fix: validateDelegationPool func in upgrades.go

dongsam commented on app/upgrades/v3rc4/testnet/upgrades.go:
```diff
@@ -0,0 +1,227 @@
+// validateDelegatePoolModuleAccount validates that the delegation pool module account is properly configured
+func validateDelegatePoolModuleAccount(ctx context.Context, ak types.AccountKeeper) error {
+	// Use hardcoded module name to avoid dependency on upgraded types
+	const delegatePoolModuleName = "epoching_delegate_pool"
```
reuse `DelegatePoolModuleName` of `/x/epoching/types/keys.go#L19`

dongsam commented on app/upgrades/v3rc4/testnet/upgrades.go:
```diff
@@ -0,0 +1,227 @@
+// validateDelegatePoolEmpty validates that the delegation pool module account has no locked funds
+func validateDelegatePoolEmpty(ctx context.Context, ak types.AccountKeeper, bk interface{}) error {
+	// Use hardcoded module name to avoid dependency on upgraded types
+	const delegatePoolModuleName = "epoching_delegate_pool"
```
ditto

dongsam commented on app/upgrades/v3rc4/testnet/upgrades.go:
```diff
@@ -0,0 +1,227 @@
+	// Direct access to keeper methods without type assertion
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+
+	currentEpoch := epochingKeeper.GetEpoch(ctx)
+	queueLength := epochingKeeper.GetCurrentQueueLength(ctx)
+
+	if queueLength > 0 {
+		return fmt.Errorf("upgrade cannot proceed with queued messages in current epoch %d (found %d queued messages) - please wait for epoch to end and messages to be processed",
```
`IncEpoch` and `InitMsgQueue` are executed in the BeginBlock of the first block of an epoch, while the upgrade is performed in the PreBlock. Is this queue length check really verifying the intended epoch?

dongsam commented on app/upgrades/v3rc4/testnet/upgrades.go:
```diff
@@ -0,0 +1,227 @@
+	// For height > 0, check if it's the first block of an epoch
+	// First block of epoch N starts at height: (N-1) * epochInterval + 1
+	// So height h is first block if: (h-1) % epochInterval == 0
+	if epochInterval == 0 {
+		return fmt.Errorf("epoch interval cannot be zero")
+	}
+	if (currentHeight-1)%epochInterval != 0 {
+		currentEpoch := ((currentHeight - 1) / epochInterval) + 1
+		nextEpochFirstBlock := (currentEpoch * epochInterval) + 1
```
Please review whether `IsFirstBlockOfNextEpoch` is already serving the same purpose, whether it can be reused, and whether the above expression is incorrectly implemented when compared with the `IsFirstBlockOfNextEpoch`.

dongsam reviewed (changes requested):
Please change the target base branch to `main` and rename the pr branch name to match its purpose
